clients/roscpp/include/ros/internal/condition_variable.h

```diff
@@ -0,0 +1,221 @@
+// http://www.boost.org/LICENSE_1_0.txt)
+// (C) Copyright 2007-10 Anthony Williams
+// (C) Copyright 2011-2012 Vicente J. Botet Escriba
+
+#ifndef ROSCPP_INTERNAL_CONDITION_VARIABLE_H
+#define ROSCPP_INTERNAL_CONDITION_VARIABLE_H
+
+#include <boost/thread/condition_variable.hpp>
+
+namespace ros {
+namespace internal {
+
+#if !defined(BOOST_THREAD_PLATFORM_PTHREAD) || \
+    defined(BOOST_THREAD_HAS_CONDATTR_SET_CLOCK_MONOTONIC) || \
+    defined(BOOST_THREAD_INTERNAL_CLOCK_IS_MONO)
+using condition_variable_monotonic = boost::condition_variable;
+
+#else
+
+class condition_variable_monotonic {
+private:
+#if defined BOOST_THREAD_PROVIDES_INTERRUPTIONS
+  pthread_mutex_t internal_mutex;
+#endif
+  pthread_cond_t cond;
+
+public:
+  condition_variable_monotonic() {
+    int res;
+#if defined BOOST_THREAD_PROVIDES_INTERRUPTIONS
+    res = pthread_mutex_init(&internal_mutex, NULL);
+    if (res)
+    {
+      boost::throw_exception(boost::thread_resource_error(res, "ros::internal::condition_variable_monotonic::condition_variable_monotonic() constructor failed in pthread_mutex_init"));
+    }
+#endif
+
+    // res = boost::detail::monotonic_pthread_cond_init(cond);
+    pthread_condattr_t attr;
+    res = pthread_condattr_init(&attr);
+    if (res == 0) {
+      pthread_condattr_setclock(&attr, CLOCK_MONOTONIC);
+      res = pthread_cond_init(&cond, &attr);
+      pthread_condattr_destroy(&attr);
+    }
+
+    if (res)
+    {
+#if defined BOOST_THREAD_PROVIDES_INTERRUPTIONS
+      BOOST_VERIFY(!pthread_mutex_destroy(&internal_mutex));
+#endif
+      boost::throw_exception(boost::thread_resource_error(res, "ros::internal::condition_variable_monotonic::condition_variable() constructor failed in detail::monotonic_pthread_cond_init"));
+    }
+  }
+
+  void notify_one() BOOST_NOEXCEPT
+  {
+#if defined BOOST_THREAD_PROVIDES_INTERRUPTIONS
+    boost::pthread::pthread_mutex_scoped_lock internal_lock(&internal_mutex);
+#endif
+    BOOST_VERIFY(!pthread_cond_signal(&cond));
+  }
+
+  void notify_all() BOOST_NOEXCEPT
+  {
+#if defined BOOST_THREAD_PROVIDES_INTERRUPTIONS
+    boost::pthread::pthread_mutex_scoped_lock internal_lock(&internal_mutex);
+#endif
+    BOOST_VERIFY(!pthread_cond_broadcast(&cond));
+  }
+
+  template <class Duration>
+  boost::cv_status wait_until(
+      boost::unique_lock<boost::mutex> &lock,
+      const boost::chrono::time_point<boost::chrono::steady_clock, Duration> &t)
+  {
+    using namespace boost::chrono;
+    typedef time_point<steady_clock, nanoseconds> nano_sys_tmpt;
+    wait_until(lock,
+               nano_sys_tmpt(ceil<nanoseconds>(t.time_since_epoch())));
+    return steady_clock::now() < t ? boost::cv_status::no_timeout : boost::cv_status::timeout;
+  }
+
+  template <class Clock, class Duration>
+  boost::cv_status wait_until(
+      boost::unique_lock<boost::mutex> &lock,
+      const boost::chrono::time_point<Clock, Duration> &t)
+  {
+    using namespace boost::chrono;
```

Comment:
these instances of `using namespace boost::chrono` in these header functions are causing build failures in my downstream packages, because it introduces potential name collisions with the same types in `std::chrono` which are included indirectly as well by the packages

Though this is probably an indicator of poor `using`-hygiene on the part of one of my dependencies, I'd still say it's a regression in Melodic. Is there any reason this code must use `using`, or could it call `boost::chrono::steady_clock` etc in these functions?

Reply:
> Is there any reason this code must use `using`, or could it call `boost::chrono::steady_clock` etc in these functions?

I don't think there is technical reason why `using` was used. Please consider to create a pull request to change it to the fully qualified types and get rid of `using namespace` in the header.

Reply:
Sounds good - I've opened https://github.com/ros/ros_comm/pull/2020 to follow up

Reply:
We ran into the same problem.